domains/client/domain-operator/src/bundle_producer_election_solver.rs

```diff
@@ -1,50 +1,101 @@
+use sp_domains::bundle_producer_election::{
+    calculate_threshold, is_below_threshold, make_transcript, BundleProducerElectionParams,
+};
+use sp_domains::{BundleProducerElectionApi, DomainId, OperatorPublicKey, ProofOfElection};
+use sp_keystore::{Keystore, KeystorePtr};
 use sp_runtime::traits::Block as BlockT;
 use sp_runtime::RuntimeAppPublic;
 use std::marker::PhantomData;
+use std::sync::Arc;
 use subspace_core_primitives::Blake2b256Hash;
+use subspace_runtime_primitives::Balance;
 
-pub(super) struct BundleProducerElectionSolver<Block, CBlock> {
+pub(super) struct BundleProducerElectionSolver<Block, CBlock, CClient> {
     keystore: KeystorePtr,
+    consensus_client: Arc<CClient>,
     _phantom_data: PhantomData<(Block, CBlock)>,
 }
 
-impl<Block, CBlock> Clone for BundleProducerElectionSolver<Block, CBlock> {
+impl<Block, CBlock, CClient> Clone for BundleProducerElectionSolver<Block, CBlock, CClient> {
     fn clone(&self) -> Self {
         Self {
             keystore: self.keystore.clone(),
+            consensus_client: self.consensus_client.clone(),
             _phantom_data: self._phantom_data,
         }
     }
 }
 
-impl<Block, CBlock> BundleProducerElectionSolver<Block, CBlock>
+impl<Block, CBlock, CClient> BundleProducerElectionSolver<Block, CBlock, CClient>
 where
     Block: BlockT,
     CBlock: BlockT,
+    CClient: ProvideRuntimeApi<CBlock>,
+    CClient::Api: BundleProducerElectionApi<CBlock, Balance>,
 {
-    pub(super) fn new(keystore: KeystorePtr) -> Self {
+    pub(super) fn new(keystore: KeystorePtr, consensus_client: Arc<CClient>) -> Self {
         Self {
             keystore,
+            consensus_client,
             _phantom_data: PhantomData,
         }
     }
 
     pub(super) fn solve_challenge(
         &self,
+        slot: Slot,
+        consensus_block_hash: CBlock::Hash,
         domain_id: DomainId,
-        _global_challenge: Blake2b256Hash,
-    ) -> sp_blockchain::Result<Option<BundleSolution<Block::Hash>>> {
-        // TODO: Implement Bundle Producer Election v2
-
-        if let Some(authority_id) = self
-            .keystore
-            .sr25519_public_keys(OperatorPublicKey::ID)
-            .into_iter()
-            .next()
+        global_challenge: Blake2b256Hash,
+    ) -> sp_blockchain::Result<Option<(ProofOfElection<Block::Hash>, OperatorPublicKey)>> {
+        let BundleProducerElectionParams {
+            current_operators,
+            total_domain_stake,
+            bundle_slot_probability,
+        } = match self
+            .consensus_client
+            .runtime_api()
+            .bundle_producer_election_params(consensus_block_hash, domain_id)?
         {
-            return Ok(Some(BundleSolution::dummy(domain_id, authority_id.into())));
+            Some(params) => params,
+            None => return Ok(None),
+        };
+
+        let vrf_sign_data = make_transcript(domain_id, &global_challenge).into_sign_data();
+
+        for operator_id in current_operators {
```

Comment:
Seems inefficient to iterate over the whole operator set, I understand this is how we do in v1, non-blocker.

Reply:
The operator set may be large, the only upper bund for this set I can think of is the `MinOperatorStake`.

Reply:
The major cost in this loop is the runtime API call, I did a quick experiment locally, the runtime API call `operator_info` takes around 15 microseconds on my machine, which means 70k operators can be optimistically iterated in one second. It's indeed not good enough if we'd like to support more than 100k operators. Let's add a TODO in this PR and resolve it in part 2 if we all agree, basically, we need to track a reverse mapping of operator_id to signing_key in the runtime, cc @vedhavyas .

Reply:
I would not not worry too much about it yet. If we really have that amazing bottleneck, we can always introduce cache storage for faster retrievals